conans/model/build_info.py

```
@@ -481,57 +482,81 @@ def relocate(el):
-                           f"lib{libname}.*dylib"]
-        dll_patterns = [f"{libname}.dll"]
-
-        def _find_matching(patterns, dirs):
-            matches = set()
-            for pattern in patterns:
-                for d in dirs:
-                    matches.update(glob.glob(f"{d}/{pattern}"))
-            if len(matches) == 1:
-                return next(iter(matches))
-
-        static_location = _find_matching(static_patterns, libdirs)
-        shared_location = _find_matching(shared_patterns, libdirs)
-        dll_location = _find_matching(dll_patterns, bindirs)
+        static_location = None
+        shared_location = None
+        dll_location = None
+        # libname is exactly the pattern, e.g., ["mylib.a"] instead of ["mylib"]
+        _, ext = os.path.splitext(libname)
+        if ext in (".lib", ".a", ".dll", ".so", ".dylib"):
+            if ext in (".lib", ".a"):
+                static_location = _find_matching(libdirs, libname)
+            elif ext in (".so", ".dylib"):
+                shared_location = _find_matching(libdirs, libname)
+            elif ext == ".dll":
+                dll_location = _find_matching(bindirs, libname)
+        else:
+            regex_static = re.compile(rf"(?:lib)?{libname}(?:[._-].+)?\.(?:a|lib)")
+            regex_shared = re.compile(rf"(?:lib)?{libname}(?:[._-].+)?\.(?:so|dylib)")
+            regex_dll = re.compile(rf"(?:.+)?{libname}(?:[._-].+)?\.dll")
+            regex_any_dll = re.compile(rf".+(?:[._-].+)?\.dll")
+            static_location = _find_matching(libdirs, regex_static)
+            shared_location = _find_matching(libdirs, regex_shared)
+            if static_location and static_location.endswith(".lib"):
+                dll_location = _find_matching(bindirs, regex_dll)
+                if not dll_location:  # any existing DLL is going to be found
+                    dll_location = _find_matching(bindirs, regex_any_dll)
```

A bit risky if more than one, I'd probably raise an error if there is more than 1 dll here.

---

Even, having more than 1, I'm returning the first lib found. So, if you want to check that there are more libs, I have to change all the `_find_matching` mechanism

---

But that is quite likely to happen, to have some folder with several dlls, and then not finding a dll that matches the current .lib, but finding another one. I think this is a bit too risky of returning too many false positives.